search-app/src/cmr/search/services/query_execution.clj

```diff
@@ -140,16 +140,75 @@
   [context query]
   [context (related-item-resolver/resolve-related-item-conditions query context)])
 
+(defn- update-facets
+  "Update the orig-facets-with-count using the info in the all-facets-with-count."
+  [orig-facets-with-count all-facets-with-count]
+  (for [title-val (map :title orig-facets-with-count)]
+    (some #(when (= title-val (:title %)) %) all-facets-with-count)))
+
+(defn- get-facets-with-count
+  "Extract out the facet part that contains title and count, amoung other things: 
+  [{:title \"t1\" :count 0} {:title \"NonExist\" :count 0} {:title \"t3\" :count 1}]
+  from the facets result.  Note: facets-result is the result from a particular field, 
+  so there is only one child under :facets :children, which is why we use first to 
+  get to the child."
+  [facets-result]
+  (-> facets-result
+      (get-in [:facets :children])
+      first
+      :children)) 
+
+(defn- get-facets-for-field-again?
```

Comment:
This way of checking of if one needs to get facets for field again is inefficient. Sometimes, it is correct to have count 0 for a facet field, so count 0 does not necessarily mean that the facet count is incorrect. I think the incorrect facet size 0 only happens when the facet field is in the query param does not rank in the list of facet fields returned, so we could make this check more accurate by only doing the additional get all facets in this case. 

We have existing function `create-facets-with-zero-matches` to handle the zero match case. I don't have time to look into all the details on how to do this the best. One laziest way of doing this is to make the `create-facets-with-zero-matches` function return a count of `-1` to make it apparent that we need to do a all facets search to get the count correct, and in the code here, we check for `-1` rather than `0`. 

Hope this makes sense.

Reply:
Looks like count 0 is not a valid value for facet field. So 0 is as good as -1. The current approach is OK.